source/validate.cpp

```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
   return SPV_SUCCESS;
 }
 
+// TODO(umar): Validate header
+spv_result_t setHeader(void* user_data, spv_endianness_t endian, uint32_t magic,
+                       uint32_t version, uint32_t generator, uint32_t id_bound,
+                       uint32_t reserved) {
+  (void)user_data;
+  (void)endian;
+  (void)version;
+  (void)generator;
+  (void)id_bound;
+  (void)reserved;
+  return SPV_SUCCESS;
+}
+
+// TODO(umar): Move this class to another file
+class validation_state_t {
+  const uint32_t* binary;
+  spv_diagnostic* diagnostic;
+  const uint32_t* program_counter;
+
+  std::set<uint32_t> declared_ids;
+  std::set<uint32_t> forward_ids;
+
+ public:
+  validation_state_t(const uint32_t* binary, spv_diagnostic* diag)
+      : binary(binary), diagnostic(diag), program_counter(binary) {}
+
+  void set_instruction_words(const uint32_t * const words) {
+    program_counter = words;
+  }
+
+  spv_result_t declare_id(uint32_t id) {
+    if (declared_ids.find(id) == std::end(declared_ids)) {
+      declared_ids.insert(id);
+    } else {
+      return diag(SPV_ERROR_INVALID_ID)
+             << "ID cannot be assigned multiple times";
+    }
+    return SPV_SUCCESS;
+  }
+
+  spv_result_t forward_declare_id(uint32_t id) {
+    forward_ids.insert(id);
+    return SPV_SUCCESS;
+  }
+
+  spv_result_t remove_if_fwd_declared(uint32_t id) {
+    forward_ids.erase(id);
+    return SPV_SUCCESS;
+  }
+
+  size_t forward_declared_id_count() { return forward_ids.size(); }
+
+  bool check_declared_id(uint32_t id) {
+    return declared_ids.find(id) != std::end(declared_ids);
+  }
+
+  const uint32_t* get_binary() const { return binary; }
+
+  const uint32_t* get_instruction_ptr() { return program_counter; }
+
+  libspirv::DiagnosticStream diag(spv_result_t error_code) {
+    return libspirv::DiagnosticStream({0, 0, (uint64_t)(program_counter - binary)},
+                            diagnostic, error_code);
+  }
+};
+
+spv_result_t SSAPass(validation_state_t& _, bool can_have_forward_declared_ids,
+                     const spv_parsed_instruction_t* inst) {
+  const uint32_t* inst_ptr = _.get_instruction_ptr();
+
+  for (int i = 0; i < inst->num_operands; i++) {
+    spv_parsed_operand_t operand = inst->operands[i];
+    spv_operand_type_t type = operand.type;
+    const uint32_t* operand_ptr = inst_ptr + operand.offset;
+
+    auto ret = SPV_ERROR_INTERNAL;
+    switch (type) {
+      case SPV_OPERAND_TYPE_RESULT_ID: {
+        _.remove_if_fwd_declared(*operand_ptr);
+        ret = _.declare_id(*operand_ptr);
+        break;
+      }
+      case SPV_OPERAND_TYPE_TYPE_ID:
+      case SPV_OPERAND_TYPE_ID: {
+        if (can_have_forward_declared_ids) {
```

Comment:
For OpBranchConditional, this is not quite right.  The condition Id must have been defined by now.  It's the true branch and false branch that could be forward references.

The general point is that whether an Id can be forward declared is a function of both opcode and argument index.